### ISSUE TITLE
 EC2: add a mechanism to disable inline security group rules, for tag-based authz scenarios

### ISSUE THREAD

**Issue by dgoetsch** [aws/aws-cdk#13256]
### Description: 

Security group constructor ALWAYS adds an egress rule to the security group declaration, even if none is defined.

The problem is the egress rule in the SecurityGroup declaration. CDK is is adding an egress rule for allowAllOutbound, even if allowAllOutbound was never called.

`addIngressRule` and `addEgressRule` already work as expected - they add new blocks to the template.  Once again: **The problem is that the constructor always adds an egress rule to the security group definition even `allowAllOutbound` is never called.**  If I delete the egress rule from the security group declaration, the entire thing works.

If no egress is defined, then it doesn't really make sense to artificially insert it into the generated template.  In our environment, artificially inserting this rule causes security group creation to fail due to IAM rules on tags.

### Reproduction Steps

Create a security group, with no Xgress rules defined until after the group is constructed:

```
 val appSecurityGroup =
      SecurityGroup.fromSecurityGroupId(
        this,
        "app-security-group-ref",
        "sg-appid"
      )

    val httpSecurityGroup =
      SecurityGroup.fromSecurityGroupId(this, "http-security-group-ref", "sg-httpid")

//Notice how there is no egress defined in the constructor
val securityGroup = SecurityGroup.Builder
      .create(this, "my-secutity-group")
      .securityGroupName("my-security-group")
      .vpc(vpc)
      .build()

    securityGroup.addIngressRule(appSecurityGroup, Port.tcp(Port))
    securityGroup.addIngressRule(httpSecurityGroup, Port.tcp(Port))
```

### What did you expect to happen?
CDK should NOT put an Egress rule in the output cdk declaration for "allowAllOutbound".

```
 secutitygroup37A50950:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: service-dependencies-dev/secutity-group
      GroupName: service-dev
      Tags:
        - Key: team
          Value: service-team
      VpcId: vpc-aaaaaa
    Metadata:
      aws:cdk:path: service-dependencies-dev/secutity-group/Resource
  secutitygroupfromservicedependenciesdevappsecuritygroupref0874CB1C63790D9CBD26:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      IpProtocol: tcp
      Description: from servicedependenciesdevappsecuritygroupref0874CB1C:6379
      FromPort: 6379
      GroupId:
        Fn::GetAtt:
          - secutitygroup37A50950
          - GroupId
      SourceSecurityGroupId: sg-aaaaaa
      ToPort: 6379
    Metadata:
      aws:cdk:path: service-dependencies-dev/secutity-group/from servicedependenciesdevappsecuritygroupref0874CB1C:6379
  secutitygroupfromservicedependenciesdevhttpsecuritygroupref4A75D33763793DE4EF63:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      IpProtocol: tcp
      Description: from servicedependenciesdevhttpsecuritygroupref4A75D337:6379
      FromPort: 6379
      GroupId:
        Fn::GetAtt:
          - secutitygroup37A50950
          - GroupId
      SourceSecurityGroupId: sg-bbbbbb
      ToPort: 6379
    Metadata:
      aws:cdk:path:service-dependencies-dev/secutity-group/from servicedependenciesdevhttpsecuritygroupref4A75D337:6379
```

### What actually happened?
CDK generated a template with an egress rule for allow all outbound in the declaration:

```
 secutitygroup37A50950:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: service-dependencies-dev/secutity-group
      GroupName: service-dev
      SecurityGroupEgress:
        - CidrIp: 0.0.0.0/0
          Description: Allow all outbound traffic by default
          IpProtocol: "-1"
      Tags:
        - Key: team
          Value: service-team
      VpcId: vpc-aaaaaa
    Metadata:
      aws:cdk:path: service-dependencies-dev/secutity-group/Resource
  secutitygroupfromservicedependenciesdevappsecuritygroupref0874CB1C63790D9CBD26:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      IpProtocol: tcp
      Description: from servicedependenciesdevappsecuritygroupref0874CB1C:6379
      FromPort: 6379
      GroupId:
        Fn::GetAtt:
          - secutitygroup37A50950
          - GroupId
      SourceSecurityGroupId: sg-aaaaaa
      ToPort: 6379
    Metadata:
      aws:cdk:path: service-dependencies-dev/secutity-group/from servicedependenciesdevappsecuritygroupref0874CB1C:6379
  secutitygroupfromservicedependenciesdevhttpsecuritygroupref4A75D33763793DE4EF63:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      IpProtocol: tcp
      Description: from servicedependenciesdevhttpsecuritygroupref4A75D337:6379
      FromPort: 6379
      GroupId:
        Fn::GetAtt:
          - secutitygroup37A50950
          - GroupId
      SourceSecurityGroupId: sg-bbbbbb
      ToPort: 6379
    Metadata:
      aws:cdk:path:service-dependencies-dev/secutity-group/from servicedependenciesdevhttpsecuritygroupref4A75D337:6379
```

### Environment

    CDK CLI Version : 1.89.0
    Framework Version: 1.89.0
    Node.js Version: v10.16.3
    OS : Docker (debian)
    Language (Version): Scala (2.13.2), Java (openjdk 11.0.9 2020-10-20
    )

### Other

referencing: https://github.com/aws/aws-cdk/issues/13204 which was closed prematurely by a maintainer who made several false assumptions without waiting for a response or additional information. Adding additional context this time around.  Please don't just close this issue without actually having a conversation first.

--- 

This is :bug: Bug Report

**Comment by rix0rrr** [comment]
Please believe me when I say it's not just about the allowAllOutbound rule. The following code:

```

//Notice how there is no egress defined in the constructor
val securityGroup = SecurityGroup.Builder
      .create(this, "my-secutity-group")
      .securityGroupName("my-security-group")
      .vpc(vpc)
      .build()

    securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(Port))
                                   ^^^^^^^ this part is different
```

Would equally lead to rules appearing in places where you don't want them to (in the `AWS::EC2::SecurityGroup` resource itself). Apart from the default egress rule, the above will have created another rule that cannot be deployed in your organization.

It's because the rule CAN be inlined, so the rule WILL be inlined.

I will keep this issue open, but my recommendations will remain the same.

Your organization has chosen to enforce rules that make it impossible to deploy certain CloudFormation templates, and CDK happens to generate CloudFormation templates of the form that your organization does not allow.

Your choices are still:

* Use escape hatches to get rid of the offending properties: https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html
* Potentially use an `Aspect` to automate this change across an entire CDK application.
* Write your own `SecurityGroup` subclass and use that instead of ours.
* File a bug report against CloudFormation and ask them to change their implementation of `AWS::EC2::SecurityGroup`, passing the tags along in the initial `CreateSecurityGroup` call.

**Comment by dgoetsch** [comment]
It is true that with cdk we are able to set up our templates in a way that won't work. I don't expect all of these to be resolved, I only expect it be `possible` to create security groups via cdk when security group access is limited. 

Limiting access to aws resources based on tags is a supported feature of IAM, and its the only way security groups support semantically scoping access to specific resources (other services allow for restricting based on name, which usually helps with this). I think ideally for us, security groups would support scoping permissions based on name, but unfortunately that is not currently a supported feature. My current organization isn't the only one that I've worked for that does this.

https://docs.aws.amazon.com/IAM/latest/UserGuide/access_tags.html
https://docs.aws.amazon.com/IAM/latest/UserGuide/tutorial_attribute-based-access-control.html
https://aws.amazon.com/blogs/security/simplify-granting-access-to-your-aws-resources-by-using-tags-on-aws-iam-users-and-roles/

I also opened an issue with the CloudFormation team before filing the original bug here, they actually thought the issue at hand could be fixed by adjusting my cdk code, (unfortunately it cannot due to how the SecurityGroup constructor works) and that conversation is still ongoing. Hopefully the CFN team is amenable to fixing the bug in security group creation.

For the short term, I will probably implement a filter on the output stacks to remove the offending rule as you suggest.

If we have a need to override `addXGressRule`, we can do that, but we don't have a need yet because those methods are currently compliant.  I don't think its outlandish for the security group to not add rules to the declaration when none are defined in the cdk code.

**Comment by rix0rrr** [comment]
> I also opened an issue with the CloudFormation team before filing the original bug here, they actually thought the issue at hand could be fixed by adjusting my cdk code,

I agree that this is a bit of a problem. There is definitely a mistake in the CloudFormation handler here, but their argument against changing anything could equally well be "well just write your CloudFormation template differently to work around our bug". The problem starts when you don't get a choice in writing your CloudFormation template because it's generated.

Alright, I am after all convinced that it's worth adding a feature to work around this CloudFormation bug, because I have a sense they will be unlikely to address this on their end. Here is my proposed fix:

----

We introduce a new context key, say:

```
@aws-cdk/aws-ec2.securityGroupNoInlineRules
```

If this context key is set to true on the SecurityGroup, we never add inline rules and only ever create separate IngressRule and EgressRule resources.

We add sections to the README and in the inline documentation of `SecurityGroup` that this context key exists, what it does, and in what situation you would use it.

This change will need some rejiggering of the default egress rule behavior, where we will have to create default additional resources that need to be removed again. Fortunately I believe there is some mechanism to either not render a resource, or remove it from the tree again.

---

We are unlikely to pick up this change ourselves anytime soon, but I would support a PR that adds the above.

**Comment by dgoetsch** [comment]
@rix0rrr I have picked this up and am working on it in my downtime.  I don't have an ETA for completion, as I'm not very experienced in this project and its not my main priority, but I understand the solution and how to implement it. I'll open a PR once I get some tests implemented.

**Comment by rix0rrr** [comment]
@dgoetsch that sounds great, thanks!

Don't hesitate to submit a draft PR, or let me know if you get stuck. The most complex parts are probably going to be around the default "don't allow anything" egress rule that we insert if `allowAllOutbound: false`, which gets removed again if a REAL rule gets inserted later.

The reason we do this is because at instantiation time we don't know yet if you're going to insert *real* rules later, and we can't have the rules table empty because then EC2 is going to assume that we meant to implicitly allow everything (I'm sure I don't need to tell you this but for completeness' sake I'm doing it anyway).

So what we do is add a fake, allow-nothing rule which we remove later. The logic there is centered around the egress rule ALWAYS being in the inline table (since it can be), but that will have to change for your PR so it will become more complicated.

**Comment by BDeus** [comment]
i think it's resolved since https://github.com/aws/aws-cdk/pull/13613

**Comment by MrArnoldPalmer** [comment]
This is in fact resolved by #13613
